File: nodekit-browser/src/types/transition.ts
import type {Expression} from "./expressions/expressions.ts";
import type {NodeId, RegisterId} from "./values.ts";

interface BaseTransition<T extends string> {
    transition_type: T
}

export interface End extends BaseTransition<'End'> {
    register_updates: Record<RegisterId, Expression>
}

export interface Go extends BaseTransition<'Go'> {
    // Leaf. Constant transition.
    to: NodeId
    register_updates: Record<RegisterId, Expression>
}

export interface IfThenElse extends BaseTransition<'IfThenElse'> {
    if: Expression
    then: Transition
    else: Transition
}

export type Transition =
    | Go
    | IfThenElse
    | End;

File: nodekit/_internal/types/transition.py
from typing import Dict

import pydantic

from nodekit._internal.types.expressions.expressions import Expression, Lit
from nodekit._internal.types.value import NodeId, RegisterId

# %%
class TransitionFunction(pydantic.BaseModel):
    ...

class Transition(pydantic.BaseModel):
    when: Expression = pydantic.Field(
        description='A Boolean-valued Expression. When it evaluates to True, this Transition is taken.',
        default=Lit(value=True),
        validate_default=True,
    )
    to: NodeId
    register_updates: Dict[RegisterId, Expression]
